### PR TITLE
[pallet-revive] write sandbox output according to the provided output buffer length

### PR DESCRIPTION
**Pull request by xermicus**
Instead of error out if the provided output buffer is smaller than what we want to write, we can just write what fits into the output buffer instead. We already write back the actual bytes written to the in-out pointer, so contracts can check it anyways.

This in turn introduces the benefit of allowing contracts to implicitly request only a portion of the returned data from calls and incantations. Which is especially beneficial for YUL as the `call` family opcodes have a return data size argument and this change removes the need to work around it in contract code.

### CODE REVIEW

File: substrate/frame/revive/fixtures/contracts/create1_with_value.rs [hunk 1]
```diff
@@ -18,8 +18,8 @@
 #![no_std]
 #![no_main]
 
-use common::{input, u256_bytes};
```
Comment (xermicus) [inline substrate/frame/revive/fixtures/contracts/create1_with_value.rs:21]: drive-by

File: substrate/frame/revive/src/wasm/runtime.rs [hunk 2]
```diff
@@ -681,18 +682,14 @@ impl<'a, E: Ext, M: ?Sized + Memory<E::T>> Runtime<'a, E, M> {
 			return Ok(())
 		}
 
-		let buf_len = buf.len() as u32;
 		let len = memory.read_u32(out_len_ptr)?;
-
-		if len < buf_len {
-			return Err(Error::<E::T>::OutputBufferTooSmall.into())
-		}
```
Comment (pgherveou) [inline substrate/frame/revive/src/wasm/runtime.rs:689]: might still be beneficial to log::debug! that when this is true

File: prdoc/pr_5743.prdoc [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    description: |
+     Instead of error out if the provided output buffer is smaller than what we want to write,
+     we can just write what fits into the output buffer instead.
+     We already write back the actual bytes written to the in-out pointer,
+     so contracts can check it anyways.
+
+     This in turn introduces the benefit of allowing contracts to implicitly request only a portion
+     of the returned data from calls and incantations.
+     Which is especially beneficial for YUL as the call family opcodes have a return data size
+     argument and this change removes the need to work around it in contract code.
```
Comment (pgherveou) [inline prdoc/pr_5743.prdoc:14]: Not too sure to get the point here

Comment (athei) [reply prdoc/pr_5743.prdoc:14]: YUL sometimes want to ignore (parts) of the output but still learn about the size. This was not possible with our code. You could even ignore everything (supply sentinel) but then also don't learn about the length or you have to read everything.

File: substrate/frame/revive/fixtures/contracts/call_diverging_out_len.rs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	let mut output_buf = [0u8; BUF_SIZE];
+	api::call(
+		uapi::CallFlags::ALLOW_REENTRY,
+		callee_address,
+		0u64,
+		0u64,
+		None,
+		&[0u8; 32],
+		&[],
+		Some(&mut &mut output_buf[..N]),
```
Comment (athei) [inline substrate/frame/revive/fixtures/contracts/call_diverging_out_len.rs:47]: You could create a binding for the `&mut` reference and assert that it has length of `N`.

**Review by pgherveou — APPROVED** [review]

File: substrate/frame/revive/fixtures/contracts/call_diverging_out_len.rs (hunk 4 above)
Comment (xermicus) [reply substrate/frame/revive/fixtures/contracts/call_diverging_out_len.rs]: Yes thats a good idea thanks! Added the assertion.

**Review by athei — APPROVED** [review]
